// @gdk/gameserver/dist/server/models/multipass/models.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ELockPolicy = exports.EPushNotificationType = exports.EGameState = void 0;
var EGameState;
(function (EGameState) {
    EGameState["PENDING"] = "PENDING";
    EGameState["ONGOING"] = "ONGOING";
    EGameState["FINISHED"] = "FINISHED";
})(EGameState = exports.EGameState || (exports.EGameState = {}));
var EPushNotificationType;
(function (EPushNotificationType) {
    EPushNotificationType["SESSION_INITIALIZED"] = "SESSION_INITIALIZED";
    EPushNotificationType["PLAYER_ENTER"] = "PLAYER_ENTER";
    EPushNotificationType["PLAYER_EXIT"] = "PLAYER_EXIT";
    EPushNotificationType["PLAYER_REMOVED"] = "PLAYER_REMOVED";
    EPushNotificationType["PAYOUT"] = "PAYOUT";
    EPushNotificationType["SESSION_CANCELLED"] = "SESSION_CANCELLED";
    EPushNotificationType["SESSION_STOPPED"] = "SESSION_STOPPED";
    EPushNotificationType["SCRIPT"] = "SCRIPT";
})(EPushNotificationType = exports.EPushNotificationType || (exports.EPushNotificationType = {}));
var ELockPolicy;
(function (ELockPolicy) {
    ELockPolicy["NONE"] = "NONE";
    ELockPolicy["ALL"] = "ALL";
    ELockPolicy["ACTION_ONLY"] = "ACTION_ONLY";
    ELockPolicy["BET_ONLY"] = "BET_ONLY";
})(ELockPolicy = exports.ELockPolicy || (exports.ELockPolicy = {}));
