// @gdk/gameserver/dist/server/models/multipass/sseChannel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SSEChannel = exports.SSECHANNEL_NOTIFICATION_TYPE = void 0;
exports.SSECHANNEL_NOTIFICATION_TYPE = "application/gdk.notification+json";
class SSEChannel {
    /**
     * Creates a new SSEChannel attached to the given room
     * @param roomId the room id attached to the notifications
     */
    static create(room) {
        const channel = new SSEChannel(room, null);
        return channel;
    }
    /**
     * Constructor
     * @param room the room reference attached to the notifications
     * @param stream the new channel stream for the person
     */
    constructor(room, stream) {
        this.privateCount = 0;
        this.broadcastCount = 0;
        this.stream = stream;
        this.room = room;
    }
    /**
     * Push notifications to player / person by writing data into
     * @param type the notification type (playerJoinded, etc)
     * @param data the data of the notification (set by the script)
     * @param sessionId the id of the room"s session
     * @param broadcast if the notification is a broadcast
     */
    push(type, data, sessionId, broadcast) {
        const payload = {
            data: {
                broadcast,
                time: Date.now(),
                sequenceNumber: broadcast ? ++this.broadcastCount : ++this.privateCount,
                roomId: this.room.id,
                sessionId,
                gameState: this.room.gameEngine.getGameState(),
                type,
                content: data,
            },
            type: exports.SSECHANNEL_NOTIFICATION_TYPE,
        };
        this.stream.write(`data: ${JSON.stringify(payload)}\n\n`);
    }
}
exports.SSEChannel = SSEChannel;
