// @gdk/gameserver/dist/server/models/motors/gameSession.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GameSessionDB = void 0;
class GameSessionDB {
    constructor() {
        this._sessions = {};
    }
    /**
     * Add session.
     * @param session
     */
    add(session) {
        this._sessions[session._id] = session;
        return session;
    }
    /**
     * Get a previously registered game session.
     * @param matchFct cmp function to find the session you want.
     * @returns the matching IGameSession.
     */
    get(matchFct) {
        return Object.values(this._sessions).find((session) => matchFct(session));
    }
    /**
     * Update session.
     * @param id
     * @param partialSession
     */
    update(id, partialSession) {
        Object.assign(this._sessions[id], partialSession);
        return this._sessions[id];
    }
    /**
     * Remove session.
     * @param id
     */
    remove(id) {
        if (id) {
            delete this._sessions[id];
        }
        else {
            this._sessions = {};
        }
    }
    /**
     * List sessions.
     */
    list() {
        return Object.values(this._sessions);
    }
}
exports.GameSessionDB = GameSessionDB;
