// @gdk/gameserver/dist/server/models/motors/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMotorsErrorMessage = exports.isMotorsError = exports.getGameData = void 0;
async function getGameData(motorsScriptManager, pageNumber, pageSize) {
    const games = motorsScriptManager.list({
        skip: pageNumber * pageSize,
        limit: pageSize,
    });
    const content = games.map((game) => {
        return {
            gameName: game.serviceName,
            gameId: game._id,
            extRef: game.serviceName,
        };
    });
    const data = {
        totalElements: content.length,
        content,
    };
    return data;
}
exports.getGameData = getGameData;
function isMotorsError(error) {
    return error.hasOwnProperty("getMotorsError");
}
exports.isMotorsError = isMotorsError;
function getMotorsErrorMessage(error) {
    if (isMotorsError(error)) {
        const motorsError = error.getMotorsError();
        return motorsError.message.join(",");
    }
    return error.message;
}
exports.getMotorsErrorMessage = getMotorsErrorMessage;
