// @falcon/bridge/.build/src/typings/messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EMessageType = exports.PopinLinkTarget = exports.PopinButtonAction = exports.EPopinLevel = void 0;
/**
 * Log level
 */
var EPopinLevel;
(function (EPopinLevel) {
    EPopinLevel["Info"] = "info";
    EPopinLevel["Warning"] = "warning";
    EPopinLevel["Error"] = "error";
})(EPopinLevel = exports.EPopinLevel || (exports.EPopinLevel = {}));
/**
 * List of actions available for a popin button.
 */
var PopinButtonAction;
(function (PopinButtonAction) {
    /**
     * Open a link.
     */
    PopinButtonAction["Link"] = "link";
    /**
     * Close the popin.
     */
    PopinButtonAction["ClosePopin"] = "closePopup";
    /**
     * Close the game through the bridge.
     */
    PopinButtonAction["CloseGame"] = "closeGame";
    /**
     * Call a callback with the popin as argument.
     */
    PopinButtonAction["Callback"] = "callback";
})(PopinButtonAction = exports.PopinButtonAction || (exports.PopinButtonAction = {}));
/**
 * Set type of link.
 */
var PopinLinkTarget;
(function (PopinLinkTarget) {
    /**
     * Open the link in an another page.
     */
    PopinLinkTarget["Blank"] = "blank";
    /**
     * Open the link in the current page.
     */
    PopinLinkTarget["Self"] = "self";
})(PopinLinkTarget = exports.PopinLinkTarget || (exports.PopinLinkTarget = {}));
/**
 * When to display the message.
 */
var EMessageType;
(function (EMessageType) {
    /**
     * Immediatly after the call of the message manager.
     */
    EMessageType["Immediate"] = "immediate";
    /**
     * At the end of a game.
     */
    EMessageType["End"] = "end";
})(EMessageType = exports.EMessageType || (exports.EMessageType = {}));
