// @gdk/gameserver/dist/server/routes/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommonITFRoutes = exports.getRoutes = void 0;
function getRoutes() {
    return [
        CommonITFRoutes.getReplayProtectionToken(),
        CommonITFRoutes.getState(),
        CommonITFRoutes.createSession(),
    ];
}
exports.getRoutes = getRoutes;
class CommonITFRoutes {
    /**
     * Get replay protection token (fake)
     */
    static getReplayProtectionToken() {
        return {
            method: "POST",
            path: `/token`,
            options: {
                handler: async (request, h) => {
                    try {
                        return h
                            .response({
                            value: "123456789",
                        })
                            .code(200);
                    }
                    catch (e) {
                        return h.response(e).code(400);
                    }
                },
            },
        };
    }
    /**
     * Retrieve state (fake)
     */
    static getState() {
        return {
            method: "GET",
            path: `/state`,
            options: {
                handler: async (request, h) => {
                    try {
                        return h
                            .response({
                            state: "1234567890",
                        })
                            .code(200);
                    }
                    catch (e) {
                        return h.response(e).code(400);
                    }
                },
            },
        };
    }
    /**
     * create session (fake)
     */
    static createSession() {
        return {
            method: "POST",
            path: `/session`,
            options: {
                handler: async (request, h) => {
                    try {
                        return h.response().code(204);
                    }
                    catch (e) {
                        return h.response(e).code(400);
                    }
                },
            },
        };
    }
}
exports.CommonITFRoutes = CommonITFRoutes;
