// @gdk/gameserver/dist/server/routes/gameConfig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GameConfigRoutes = void 0;
const swagger_1 = require("./swagger");
class GameConfigRoutes {
    /**
     * Constructor
     * @param gameConfig
     */
    constructor(gameConfig) {
        this.gameConfig = gameConfig;
    }
    /**
     * Retrieve routes
     */
    getRoutes() {
        return [
            {
                method: "GET",
                path: `/gameConfig`,
                options: {
                    handler: async (request, h) => {
                        try {
                            const conf = this.gameConfig.getConfig();
                            return h.response(conf).code(200);
                        }
                        catch (e) {
                            return h.response(e).code(400);
                        }
                    },
                    description: "Get game config",
                    notes: `Retrieve json configuration for the game`,
                    plugins: {
                        "hapi-swagger": {
                            responses: {
                                200: {
                                    description: "Success",
                                    schema: swagger_1.Schema.shared.gameConfig,
                                },
                                400: {
                                    description: "Bad Request",
                                    schema: swagger_1.Schema.shared.error,
                                },
                            },
                        },
                    },
                    tags: [swagger_1.TAGS.API],
                },
            },
        ];
    }
}
exports.GameConfigRoutes = GameConfigRoutes;
