// @falcon/bridge/.build/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.game = exports.EPushTransportMechanism = exports.EItgPrizeType = exports.ECurrencyPurseType = exports.loadFile = exports.loadJSON = exports.loadScript = exports.injectScript = exports.uid = exports.getQueryParameters = exports.ChannelProxy = exports.EGameMode = exports.Bridge = void 0;
var falcon_game_bridge_1 = require("./falcon-game-bridge");
var falcon_game_bridge_2 = require("./falcon-game-bridge");
Object.defineProperty(exports, "Bridge", { enumerable: true, get: function () { return falcon_game_bridge_2.Bridge; } });
Object.defineProperty(exports, "EGameMode", { enumerable: true, get: function () { return falcon_game_bridge_2.EGameMode; } });
var channel_1 = require("./channel");
Object.defineProperty(exports, "ChannelProxy", { enumerable: true, get: function () { return channel_1.ChannelProxy; } });
var tools_1 = require("./tools/tools");
Object.defineProperty(exports, "getQueryParameters", { enumerable: true, get: function () { return tools_1.getQueryParameters; } });
Object.defineProperty(exports, "uid", { enumerable: true, get: function () { return tools_1.uid; } });
Object.defineProperty(exports, "injectScript", { enumerable: true, get: function () { return tools_1.injectScript; } });
Object.defineProperty(exports, "loadScript", { enumerable: true, get: function () { return tools_1.loadScript; } });
Object.defineProperty(exports, "loadJSON", { enumerable: true, get: function () { return tools_1.loadJSON; } });
Object.defineProperty(exports, "loadFile", { enumerable: true, get: function () { return tools_1.loadFile; } });
var common_1 = require("./typings/common");
Object.defineProperty(exports, "ECurrencyPurseType", { enumerable: true, get: function () { return common_1.ECurrencyPurseType; } });
var platform_1 = require("./typings/platform");
Object.defineProperty(exports, "EItgPrizeType", { enumerable: true, get: function () { return platform_1.EItgPrizeType; } });
Object.defineProperty(exports, "EPushTransportMechanism", { enumerable: true, get: function () { return platform_1.EPushTransportMechanism; } });
__exportStar(require("./typings/mtr"), exports);
exports.game = { bridge: falcon_game_bridge_1.Bridge };
exports.default = "test";
