// @gdk/gameserver/dist/server/models/motors/tools/histoRenderer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const gameModel_1 = __importDefault(require("./gameModel"));
const toJs_1 = require("../groovy/toJs");
class HistoRenderer {
    /**
     * Render game session.
     */
    static async render(gameSession) {
        const gameModel = await gameModel_1.default.get(gameSession.manifestFilePath);
        const renderer = new HistoRenderer(gameModel);
        return renderer.execute(gameSession);
    }
    constructor(gameModel) {
        this.xmlSnippet = gameModel.histoRenderFunction;
        this.constants = gameModel.constants;
    }
    /**
     * Execute HistoRenderFunction.
     * @param gameSession
     */
    async execute(gameSession) {
        const vm = toJs_1.GroovyToJS.create(this.xmlSnippet, "HistoRenderFunction");
        const params = {
            events: gameSession.events,
            constants: this.constants,
            betCurrency: gameSession.amountCurrency,
        };
        return vm.execute(params);
    }
}
exports.default = HistoRenderer;
